### PR TITLE
scopegen: add scopegen plugin to generate OAUth2 scopes

### PR DESCRIPTION
**Pull request by favadi**
Currently, the list of supported languages are:

- json
- go

### CODE REVIEW

File: scopegen/parser/parser.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+		methods = make(map[string][]string)
+		result  []Method
+	)
+
+	if securitySchemeName == "" {
+		return nil, nil
+	}
+
+	for _, service := range file.GetService() {
+		for _, method := range service.GetMethod() {
```
Comment (mgfcmartinez) [inline scopegen/parser/parser.go]: nit: this block is meaty enough to warrant its own function IMO (`extractMethodsFromService`)

File: scopegen/generate/go.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	"io"
+
+	"github.com/gunk/gunk/scopegen/parser"
+)
+
+var goTmpl = template.Must(template.New("").Parse(`// Code generated by "scopegen"; DO NOT EDIT.
+package {{ .Package }}
+
+type AuthScope string
+{{ range $scope := .Scopes }}
```
[the remark this answers is not in the document]
Comment (favadi) [reply scopegen/generate/go.go]: fixed in latest commit.

**Comment by favadi** [comment]
> just nits, LGTM. perhaps testcases for handling undefined scopes/params?

@mgfcmartinez fixed all your comments, also add test case for undefined scope in scopegen_undefined_scope.txt.

File: scopegen/scopegen.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+			})
+		case goLangName:
+			var buf = bytes.NewBuffer(nil)
+			if err = generate.Go(buf, parsed); err != nil {
+				return nil, fmt.Errorf("failed to generate scopes: lang=%s err=%s", lang, err.Error())
+			}
+			resp.File = append(resp.File, &plugin_go.CodeGeneratorResponse_File{
+				Name:    proto.String(filepath.Join(baseDir, "all.scopes.go")),
+				Content: proto.String(buf.String()),
+			})
```
Comment (mvdan) [inline scopegen/scopegen.go]: nit: these four lines are practically the same in all cases, minus the file name; but that's just `"all.scopes."+lang`. Why not move it out of the switch?

File: scopegen/parser/parser.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	swagger := fileExt.(*options.Swagger)
+
+	if swagger.GetSecurityDefinitions() == nil {
+		return "", nil, nil
+	}
+	for securitySchemeName, securityScheme := range swagger.GetSecurityDefinitions().GetSecurity() {
+		if securityScheme.Type == options.SecurityScheme_TYPE_OAUTH2 {
+			if securityScheme.Scopes != nil {
```
Comment (mvdan) [inline scopegen/parser/parser.go]: invert the `ifs` and continue early to unindent the code

File: scopegen/scopegen.go (hunk 3 above)
Comment (favadi) [reply scopegen/scopegen.go]: refactored and moved out of the switch.

**Pull request merged**